### ISSUE TITLE
Breaking changes with publish tasks

### ISSUE THREAD

**Issue by pascalberger** [Microsoft/vso-agent-tasks#986]
In the last two sprints the behavior of the copy & publish task has twice changed. It's quite a paint to have working build definitions suddenly breaking because of some change in the task, especially if it was nowhere announced.
The problem was that the task handled the copying of subfolders differently resulting in different structure in the artifacts.

Whats the policy regarding breaking changes? When can they occur and how can I be informed about them?

**Comment by bryanmacfarlane** [comment]
The intent was not to change the behavior.  This was a simple performance fix with a subtle change in behavior we didn't notice.  So, just a bug.  When code changes, it can always happen.  We hear all the time, we want x, y, and z added - but make sure nothing changes :)  The issue we fixed was duplicated copy (more time spent preparing to upload rather than uploading)

We patched the task.

BTW, tasks are fully versioned in the backend.  At one point, we had a UI which allowed you to lock to a version of a task and choose when to move forward or you could choose latest.  That has the downside of you having to constantly go edit your defs to get new features and bug fixes.  The feature was hidden/cut right before our preview release but we could add it back.

>> Whats the policy regarding breaking changes?

the policy is to not have breaking changes if at all policy.  we go through great pains and testing to confirm that.  

>> When can they occur and how can I be informed about them?

As stated, the policy is to not break previous behavior.  If for some reason we, we would try to put feedback right in the code/task in the form of a warning/log output.

**Comment by pascalberger** [comment]
Thanks @bryanmacfarlane. I completely understand that bugs can happen. It's just a pain when suddenly all builds/deployments stop working and you have to figure out what has changed (another current story is that since last sprint GitVersion results in wrong version numbers).

IMHO it would be a nice feature if tasks by default are on the latest version but if there is a possibility to lock it back to a previous version. With this one would have the possibility to quickly switch back and have everything working again and time to adopt to changes. Another story are changes on the agent where I don't know if it is realistic to have several version supported in parallel.

**Comment by bryanmacfarlane** [comment]
Yes, when we had it implemented, latest was the default.

I like the scenario that you point out - a bug - use latest.  I'll likely use that in bringing the feature back.  Having a 'previous' option would be cool
